pkg/controller/build/build_controller.go

```diff
@@ -1160,62 +1160,129 @@ func (ctrl *Controller) validatePullSecret(name string) (*corev1.Secret, error)
+		return nil, err
+	}
+
+	// If neither the secret from the API nor our canonicalization attempt are
+	// identified as canonicalized secrets, that means it does not need to be
+	// canonicalized. So we can return it as-is.
+	if !isCanonicalizedSecret(canonicalized) && !isCanonicalizedSecret(secret) {
+		return secret, nil
+	}
+
+	// If the secret from the API server is not canonicalized, but it is
+	// identified as canonicalized by our canonicalization attempt, that means
+	// the canonicalized secret needs to be created.
+	if isCanonicalizedSecret(canonicalized) && !isCanonicalizedSecret(secret) {
+		klog.Infof("Found legacy-style secret %s, canonicalizing as %s", secret.Name, canonicalized.Name)
+		return ctrl.handleCanonicalizedPullSecret(canonicalized)
 	}
 
-	return secret, nil
+	// If we have a canonicalized secret, get the original secret name from the
+	// label, then retry validation with the original secret. This will cause the
+	// canonicalized secret to be updated if the original secret has changed.
+	return ctrl.validatePullSecret(canonicalized.Labels[originalSecretNameLabel])
 }
 
 // Attempt to create a canonicalized pull secret. If the secret already exsits, we should update it.
 func (ctrl *Controller) handleCanonicalizedPullSecret(secret *corev1.Secret) (*corev1.Secret, error) {
+	if err := validateCanonicalizedSecret(secret); err != nil {
+		return nil, fmt.Errorf("could not handle canonicalized secret: %w", err)
+	}
+
 	out, err := ctrl.kubeclient.CoreV1().Secrets(ctrlcommon.MCONamespace).Get(context.TODO(), secret.Name, metav1.GetOptions{})
-	if err != nil && !k8serrors.IsNotFound(err) {
-		return nil, fmt.Errorf("could not get canonical secret %q: %w", secret.Name, err)
+	// We found a canonicalized secret.
+	if err == nil {
+		// Check if the canonical secret from the API server matches the secret we
+		// got. If they match, we're done.
+		if bytes.Equal(secret.Data[corev1.DockerConfigJsonKey], out.Data[corev1.DockerConfigJsonKey]) && hasCanonicalizedSecretLabels(secret) {
+			klog.Infof("Canonical secret %q up-to-date", secret.Name)
+			return out, nil
+		}
+
+		// If they do not match, we need to do an update.
+		return ctrl.updateCanonicalizedSecret(secret)
 	}
 
 	// We don't have a canonical secret, so lets create one.
 	if k8serrors.IsNotFound(err) {
-		out, err = ctrl.kubeclient.CoreV1().Secrets(ctrlcommon.MCONamespace).Create(context.TODO(), secret, metav1.CreateOptions{})
-		if err != nil {
-			return nil, fmt.Errorf("could not create canonical secret %q: %w", secret.Name, err)
-		}
+		return ctrl.createCanonicalizedSecret(secret)
+	}
 
-		klog.Infof("Created canonical secret %s", secret.Name)
-		return out, nil
+	return nil, fmt.Errorf("could not get canonicalized secret %q: %w", secret.Name, err)
+}
+
+// Creates a canonicalized secret.
+func (ctrl *Controller) createCanonicalizedSecret(secret *corev1.Secret) (*corev1.Secret, error) {
+	if err := validateCanonicalizedSecret(secret); err != nil {
```

Comment:
This call seems redundant as the calling function would have run validateCanonicalizedSecret()

Reply:
It may be redundant, though I added it purely as a guard clause because there is no way to distinguish a non-canonicalized `corev1.Secret` and a canonicalized one. In the future, introducing another type that represents canonicalized secrets might be a better way forward.

Reply:
fair enough, that makes sense to me!